File: app/modules/ytdlp/android/src/main/java/expo/modules/ytdlp/Url.kt
package expo.modules.ytdlp

import android.net.Uri

private val HAS_SCHEME = Regex("^[a-zA-Z][a-zA-Z\\d+.-]*://")

/**
 * Mirrors normalizeUrl() in server/src/url.ts. Adds "https://" to a schemeless link like
 * "youtube.com/watch?v=x".
 */
internal fun normalizeUrl(url: String): String {
    val trimmed = url.trim()
    return if (HAS_SCHEME.containsMatchIn(trimmed)) trimmed else "https://$trimmed"
}

/** Whether this URI uses a scheme yt-dlp/the app actually supports (as opposed to e.g. ftp:, file:). */
internal fun Uri.isHttpOrHttps(): Boolean {
    val normalizedScheme = scheme?.lowercase()
    return normalizedScheme == "http" || normalizedScheme == "https"
}

File: app/server/url.ts
const HAS_SCHEME = /^[a-z][a-z\d+.-]*:\/\//i;

/** Adds a "https://" prefix to a schemeless link like "youtube.com/watch?v=x" so it can be parsed as a URL. */
export function normalizeUrl(input: string): string {
  const trimmed = input.trim();
  return HAS_SCHEME.test(trimmed) ? trimmed : `https://${trimmed}`;
}
